XenModel/Actions/HealthCheck/XenServerHealthCheckUpload.cs

```diff
@@ -175,62 +159,60 @@ public string UploadZip(string fileName, string caseNumber, System.Threading.Can
             long size = fileInfo.Length;
 
             // Fetch the upload UUID from CIS server.
-            string uploadUuid = InitiateUpload(Path.GetFileName(fileName), size, caseNumber);
+            string uploadUuid = InitiateUpload(Path.GetFileName(fileName), size, caseNumber, cancel);
             if (string.IsNullOrEmpty(uploadUuid))
             {
                 log.ErrorFormat("Cannot fetch the upload UUID from CIS server");
                 return "";
             }
 
             // Start to upload zip file.
-            log.InfoFormat("Upload server returns Upload ID: {0}", uploadUuid);
-
-            long offset = 0;
-            while (offset < size)
+            log.InfoFormat("Upload server returned Upload ID: {0}", uploadUuid);
+            using (var source = File.Open(fileName, FileMode.Open))
             {
-                StringBuilder url = new StringBuilder(UPLOAD_URL + "upload_raw_chunk/?id=" + uploadUuid);
-                url.Append(String.Format("&offset={0}", offset));
-                long remainingSize = size - offset;
-                long thisChunkSize = (remainingSize > CHUNK_SIZE) ? CHUNK_SIZE : remainingSize;
-
-                try
+                long offset = 0;
+                while (offset < size)
                 {
-                    for (int i = 0; i < 3; i++)
-                    {
-                        if (cancel.IsCancellationRequested)
-                        {
-                            log.Info("Upload cancelled");
-                            return "";
-                        }
+                    var url = string.Format(UPLOAD_CHUNK_STRING, uploadUrl, uploadUuid, offset);
+                    long remainingSize = size - offset;
+                    long thisChunkSize = remainingSize > CHUNK_SIZE ? CHUNK_SIZE : remainingSize;
 
-                        if (UploadChunk(url.ToString(), fileName, offset, thisChunkSize, uploadToken))
-                        {
-                            // This chunk is successfully uploaded
-                            offset += thisChunkSize;
-                            break;
-                        }
-
-                        // Fail to upload the chunk for 3 times so fail the bundle upload.
-                        if (i == 2)
+                    try
+                    {
+                        for (int i = 0; i < 3; i++)

```

Comment:
Let's put 3 in a constant and use it further down, so as to have one place to change if we update the number of retries.